### PR TITLE
Create documentation of `config.yaml`

### LINKED ISSUE #129: Create documentation of `config.yaml`
Create a page under "References" that goes over `config.yaml`.

### PR DESCRIPTION
**Pull request by sujuka99**
Closes #129

### CODE REVIEW

File: examples/bakdata/atm-fraud-detection/config.yaml [hunk 1]
```diff
@@ -11,3 +11,5 @@ schema_registry_url: "http://k8kafka-cp-schema-registry.kpops.svc.cluster.local:
 kafka_rest_host: "http://localhost:8082"
 
 kafka_connect_host: "http://localhost:8083"
+
+defaults_path: ./defaults
```
Comment (disrupted) [inline examples/bakdata/atm-fraud-detection/config.yaml:15]: if we're setting it in the config.yaml, we can remove it here https://github.com/bakdata/kpops/blob/main/docs/docs/user/examples/atm-fraud-pipeline.md

Comment (disrupted) [reply examples/bakdata/atm-fraud-detection/config.yaml:15]: good call btw 👍🏻

File: docs/docs/resources/pipeline_config/config.yaml [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+# Global flags for Helm
+helm_config:
+  # Set the name of the kubeconfig context. (--kube-context)
+  context: name
+  # Run Helm in Debug mode.
+  debug: false
+# Configure Helm Diff
+helm_diff_config: 
+  # Enable Helm Diff.
+  enable: false
```
Comment (disrupted) [inline docs/docs/resources/pipeline_config/config.yaml:45]: we should change the default to true imo

File: docs/docs/resources/pipeline_config/config.yaml [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+# The path to the folder containing the defaults file and the environment
+# defaults files.
+# REQUIRED
+defaults_path: /path/to/defaults
+# The environment you want to generate and deploy the pipeline to. Suffix your
+# environment files with this value (e.g. defaults_development.yaml for
+# environment=development).
+# REQUIRED
+environment: development
+# The kafka broker address.
```
Comment (disrupted) [inline docs/docs/resources/pipeline_config/config.yaml:12]:
```suggestion
# The Kafka broker address.
```

File: docs/docs/resources/pipeline_config/config.yaml (hunk 2 above)
Comment (sujuka99) [reply docs/docs/resources/pipeline_config/config.yaml:45]: Agreed, should I do it?

File: docs/docs/resources/pipeline_config/config.yaml [hunk 4]
```diff
@@ -0,0 +1,55 @@
+# CONFIGURATION
+#
+# The path to the folder containing the defaults file and the environment
```
Comment (disrupted) [inline docs/docs/resources/pipeline_config/config.yaml:3]:
```suggestion
# The path to the folder containing the defaults.yaml file and the environment
```

File: docs/docs/resources/pipeline_config/config.yaml [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+# The environment you want to generate and deploy the pipeline to. Suffix your
+# environment files with this value (e.g. defaults_development.yaml for
+# environment=development).
+# REQUIRED
+environment: development
+# The kafka broker address.
+# REQUIRED
+broker: "http://localhost:9092"
+# The name of the defaults file and the prefix of the defaults environment file.
+defaults_filename_prefix: defaults
```
Comment (disrupted) [inline docs/docs/resources/pipeline_config/config.yaml:16]: I had no idea you can customize this 😄

File: docs/docs/resources/pipeline_config/config.yaml [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+# Base class for settings, allowing values to be overridden by environment
+# variables. This is useful in production for secrets you do not wish to save in
+# code, it plays nicely with docker(-compose), Heroku and any 12 factor app
+# design.
```
Comment (disrupted) [inline docs/docs/resources/pipeline_config/config.yaml:20]: where did you get this info from?

File: docs/docs/resources/pipeline_config/config.yaml [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+topic_name_config: 
+  # Configures the value for the variable ${output_topic_name}
+  default_output_topic_name: ${pipeline_name}-${component_name}
+  # Configures the value for the variable ${error_topic_name}
+  default_error_topic_name: ${pipeline_name}-${component_name}-error
+# Address to the rest proxy REST API.
```
Comment (disrupted) [inline docs/docs/resources/pipeline_config/config.yaml:26]:
```suggestion
# Address of the Kafka REST Proxy.
```

File: docs/docs/resources/pipeline_config/config.yaml [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+kafka_rest_host: "http://localhost:8082"
+# Address to the kafka connect REST API.
```
Comment (disrupted) [inline docs/docs/resources/pipeline_config/config.yaml:28]:
```suggestion
# Address of Kafka Connect.
```

File: docs/docs/resources/pipeline_config/config.yaml [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+kafka_connect_host: "http://localhost:8083"
+# The timeout in seconds that specifies when actions like deletion or deploy
+# timeout.
+timeout: 300
+# Flag for `helm upgrade --install`.
+# Create the release namespace if not present.
+create_namespace: false
```
Comment (disrupted) [inline docs/docs/resources/pipeline_config/config.yaml:35]: weird, why is this here and not under `helm_config`

File: docs/docs/resources/pipeline_config/config.yaml [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+  # Set of keys that should not be checked.
+  ignore: 
+    - name
+    - imageTag
+# Whether to retain clean up jobs in the cluster or uninstall the, after
+# completion.
+retain_clean_jobs: false
+# The URL to schema registry.
+schema_registry_url: "http://localhost:8081"
+
```
Comment (disrupted) [inline docs/docs/resources/pipeline_config/config.yaml:55]:
```suggestion
# Address of the Schema Registry.
schema_registry_url: "http://localhost:8081"
```
move this up before Kafka REST Proxy

**Review by disrupted — CHANGES_REQUESTED** [review]

File: docs/docs/resources/pipeline_config/config.yaml (hunk 6 above)
Comment (sujuka99) [reply docs/docs/resources/pipeline_config/config.yaml:20]: Description of the fields in the schema :)

File: docs/docs/resources/pipeline_config/config.yaml (hunk 9 above)
Comment (sujuka99) [reply docs/docs/resources/pipeline_config/config.yaml:35]: Yeah, it's a bit weird

File: examples/bakdata/atm-fraud-detection/config.yaml (hunk 1 above)
Comment (sujuka99) [reply examples/bakdata/atm-fraud-detection/config.yaml:15]: Thanks, the schema says it is required in `config.yaml`. Does it get overridden if you provide a different path from the CLI?

File: docs/docs/resources/pipeline_config/config.yaml (hunk 5 above)
Comment (sujuka99) [reply docs/docs/resources/pipeline_config/config.yaml:16]: I'm so grateful for the schema, this would've taken soo long without it!

File: examples/bakdata/atm-fraud-detection/config.yaml (hunk 1 above)
Comment (sujuka99) [reply examples/bakdata/atm-fraud-detection/config.yaml:15]: Fixed the example doc.

File: docs/docs/resources/pipeline_config/config.yaml (hunk 2 above)
Comment (sujuka99) [reply docs/docs/resources/pipeline_config/config.yaml]: Well, I did. Lmk if I should revert it.

File: docs/docs/resources/pipeline_config/config.yaml (hunk 3 above)
Comment (sujuka99) [reply docs/docs/resources/pipeline_config/config.yaml]: Done

File: docs/docs/resources/pipeline_config/config.yaml (hunk 7 above)
Comment (sujuka99) [reply docs/docs/resources/pipeline_config/config.yaml]: done

File: docs/docs/resources/pipeline_config/config.yaml (hunk 8 above)
Comment (sujuka99) [reply docs/docs/resources/pipeline_config/config.yaml]: Done

File: docs/docs/resources/pipeline_config/config.yaml (hunk 10 above)
Comment (sujuka99) [reply docs/docs/resources/pipeline_config/config.yaml]: Done

File: docs/docs/user/examples/atm-fraud-pipeline.md [hunk 11]
```diff
@@ -79,7 +79,6 @@ kubectl port-forward --namespace kpops service/k8kafka-cp-kafka-connect 8083:808
     ```shell
     poetry run kpops deploy ./examples/bakdata/atm-fraud-detection/pipeline.yaml \
     --pipeline-base-dir ./examples \
-    --defaults ./examples/bakdata/atm-fraud-detection/defaults \
```
Comment (disrupted) [inline docs/docs/user/examples/atm-fraud-pipeline.md:82]: I just tried running this command without `--defaults` and this is actually broken right now. most likely because of https://github.com/bakdata/kpops/pull/138/files#r1123381257

File: kpops/cli/pipeline_config.py [hunk 12]
```diff
@@ -45,25 +45,31 @@ class PipelineConfig(BaseSettings):
...
         example="http://localhost:8082",
-        description="Address to the rest proxy REST API.",
+        description="Address of the Kafka REST Proxy.",
     )
     kafka_connect_host: str | None = Field(
         default=None,
         env=f"{ENV_PREFIX}CONNECT_HOST",
         example="http://localhost:8083",
-        description="Address to the kafka connect REST API.",
+        description="# Address of Kafka Connect.",
```
Comment (disrupted) [inline kpops/cli/pipeline_config.py:64]:
```suggestion
        description="Address of Kafka Connect.",
```
typo

File: docs/docs/user/references/config.md [hunk 13]
```diff
@@ -1 +1,15 @@
 # Configuration
+
+KPOps reads most of a [pipeline](/user/getting-started/quick-start/#deploying-the-word-count-pipeline)'s configuration that is not related to its [components](./components.md) from [`config.yaml`](/user/references/config).
```
Comment (disrupted) [inline docs/docs/user/references/config.md:3]:
```suggestion
KPOps reads most of a [pipeline](/user/getting-started/quick-start/#deploying-the-word-count-pipeline)'s configuration that is not related to its [components](./components.md) from [`config.yaml`](/user/references/config.md).
```
necessary?

File: docs/docs/resources/pipeline_config/config.yaml (hunk 6 above)
Comment (disrupted) [reply docs/docs/resources/pipeline_config/config.yaml:20]: I can't make sense of this. what does it even belong to? do you know? it definitely doesn't belong to `topic_name_config`

File: kpops/cli/pipeline_config.py [hunk 14]
```diff
@@ -23,7 +23,7 @@ class TopicNameConfig(BaseSettings):
 class PipelineConfig(BaseSettings):
```
Comment (disrupted) [inline kpops/cli/pipeline_config.py:23]: please run `python gen_schema.py` after changing the descriptions :)

File: kpops/cli/pipeline_config.py [hunk 15]
```diff
@@ -45,25 +45,31 @@ class PipelineConfig(BaseSettings):
...
+        description="# Address of Kafka Connect.",
     )
     timeout: int = Field(
         default=300,
         env=f"{ENV_PREFIX}TIMEOUT",
         description="The timeout in seconds that specifies when actions like deletion or deploy timeout.",
     )
 
-    create_namespace: bool = Field(default=False)
+    create_namespace: bool = Field(default=False)  # Move under helm_config?
```
Comment (disrupted) [inline kpops/cli/pipeline_config.py:72]: please move it :)

File: docs/docs/user/references/config.md (hunk 13 above)
Comment (sujuka99) [reply docs/docs/user/references/config.md:3]: I suppose at least a little explanation on what the file is used for could be beneficial for a complete newbie?

Comment (disrupted) [reply docs/docs/user/references/config.md:3]: I was referring to the `.md` file ending because it is used for the other link

File: examples/bakdata/atm-fraud-detection/config.yaml (hunk 1 above)
Comment (sujuka99) [reply examples/bakdata/atm-fraud-detection/config.yaml:15]: Should I assign it to v1.0

File: kpops/cli/pipeline_config.py (hunk 15 above)
Comment (sujuka99) [reply kpops/cli/pipeline_config.py:72]: I actually think that we should keep it there. HelmConfig contains only global flags, which `--create-namespace` isn't. 

Kinda unrelated, but it would be cool to have a `namespace` flag in `HelmConfig` instead of having to define it in `defaults.yaml`.

File: docs/docs/resources/pipeline_config/config.yaml (hunk 9 above)
Comment (sujuka99) [reply docs/docs/resources/pipeline_config/config.yaml:37]: https://github.com/bakdata/kpops/pull/138#discussion_r1124262751

File: kpops/cli/pipeline_config.py (hunk 14 above)
Comment (sujuka99) [reply kpops/cli/pipeline_config.py:27]: done

File: docs/docs/user/references/config.md (hunk 13 above)
Comment (sujuka99) [reply docs/docs/user/references/config.md]: Ah, no it's not needed. I opted for removing it from the other one as well.

File: kpops/cli/pipeline_config.py (hunk 12 above)
Comment (sujuka99) [reply kpops/cli/pipeline_config.py]: Fixed

File: docs/docs/user/examples/atm-fraud-pipeline.md (hunk 11 above)
Comment (sujuka99) [reply docs/docs/user/examples/atm-fraud-pipeline.md]: Reverted,

Could be once again removed in #144

File: docs/docs/resources/pipeline_config/config.yaml (hunk 6 above)
Comment (sujuka99) [reply docs/docs/resources/pipeline_config/config.yaml]: Belongs to `BaseSettings`, I added docstrings to the classes to override it (and to describe them a bit).

File: examples/bakdata/atm-fraud-detection/config.yaml (hunk 1 above)
Comment (disrupted) [reply examples/bakdata/atm-fraud-detection/config.yaml:15]: yes please. :) this bug needs to be fixed

File: docs/docs/resources/pipeline_config/config.yaml (hunk 6 above)
Comment (disrupted) [reply docs/docs/resources/pipeline_config/config.yaml]: great, thank you!

File: docs/docs/user/references/config.md [hunk 16]
```diff
@@ -1 +1,15 @@
 # Configuration
+
+KPOps reads most of a [pipeline](/user/getting-started/quick-start/#deploying-the-word-count-pipeline)'s configuration that is unrelated to its [components](./components) from [`config.yaml`](/user/references/config).
+
+Consider enabling [KPOps' editor integration](/user/references/editor-integration) feature to enjoy the benefits of autocompletion and validation when configuring your pipeline.
+
+To learn about any of the available settings, take a look at the example below.
+
+??? "`config.yaml`"
```
Comment (disrupted) [inline docs/docs/user/references/config.md:9]: didn't you have ???+ before to have the panel open automatically?

Comment (sujuka99) [reply docs/docs/user/references/config.md:9]: It was on the Editor Integration page. Here I dislike how it looks because the file is too long to fit on the screen in its entirety. Do you think I should expand it by default?

Comment (disrupted) [reply docs/docs/user/references/config.md:9]: ok, makes sende, then it's fine as it is.

**Review by disrupted — APPROVED** [review]
LGTM
